sdk/iot/core/src/az_iot_hub_client.c

```diff
@@ -1,12 +1,133 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // SPDX-License-Identifier: MIT
 
+#include <stdint.h>
+
 #include "az_iot_hub_client.h"
+#include <az_precondition.h>
+#include <az_result.h>
 #include <az_span.h>
 
 #include <_az_cfg.h>
 
+static const uint8_t hub_client_forward_slash = '/';
+static const uint8_t hub_client_param_separator = '&';
+static const uint8_t hub_client_null_terminate = '\0';
+
+static const az_span hub_client_api_version = AZ_SPAN_LITERAL_FROM_STR("?api-version=2018-06-30");
+
 AZ_NODISCARD az_iot_hub_client_options az_iot_hub_client_options_default()
 {
   return (az_iot_hub_client_options){ .module_id = AZ_SPAN_NULL, .user_agent = AZ_SPAN_NULL };
 }
+
+AZ_NODISCARD az_result az_iot_hub_client_init(
+    az_iot_hub_client* client,
+    az_span iot_hub_hostname,
+    az_span device_id,
+    az_iot_hub_client_options const* options)
+{
+  AZ_PRECONDITION_NOT_NULL(client);
+  AZ_PRECONDITION_VALID_SPAN(iot_hub_hostname, 1, false);
+  AZ_PRECONDITION_VALID_SPAN(device_id, 1, false);
+  AZ_PRECONDITION_NOT_NULL(options);
```

Comment:
@CIPop Should `NULL` be allowed here or should the behavior be that the user either uses a custom set (which could be all options `AZ_SPAN_NULL`) or uses default. But in either case options isn't `NULL`.

Reply:
[Current header says it can](https://github.com/Azure/azure-sdk-for-c/blob/95ff0e13dcaf5a6d70014037b1e652fc73b2ef74/sdk/iot/core/inc/az_iot_hub_client.h#L61) but in the storage SDK `NULL` isn't allowed. [Link here](https://github.com/Azure/azure-sdk-for-c/blob/95ff0e13dcaf5a6d70014037b1e652fc73b2ef74/sdk/storage/blobs/src/az_storage_blobs_blob_client.c#L61)

Reply:
Please follow the guidelines as other SDKs may still need to be updated to follow the latest docs.
https://azure.github.io/azure-sdk/clang_design.html#functions-with-many-parameters

Reply:
NULL should work: the implementation will assume (call) defaults.